src/config/format.rs

```diff
@@ -0,0 +1,261 @@
+//! Support for loading configs from multiple formats.
+
+#![deny(missing_docs, missing_debug_implementations)]
+
+use serde::de;
+use std::path::Path;
+
+/// The format used to represent the configuration data.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub enum Format {
+    /// The format could not be determined.
+    Unknown,
+    /// TOML format is used.
+    TOML,
+    /// JSON format is used.
+    JSON,
+    /// YAML format is used.
+    YAML,
+}
+
+impl Default for Format {
+    fn default() -> Self {
+        Format::TOML
+    }
+}
+
+impl Format {
+    /// Returns the format as is unless it's unknown.
+    /// If the format is unknown - executes the function and return the result
+    /// of the function.
+    pub fn known_or<F>(self, f: F) -> Self
```

Comment:
Why not store `Option<Format>` and use `unwrap_or`?

Reply:
I thought about this too after the last refactoring. I think we can switch to `Option<Format>`. The downside is it'd be a bit more verbose and thus harder to read, the upside is it's going to be more familiar to people.
It'll also expose the composition of the `Format` (i.e. allow it to be passed without `Option<T>` wrapper) which I don't want here - my thinking was that we want to encapsulate all the logic of the final fallback at the `deserialize` call (i.e. scope it to the `format` mod, and make sure it doesn't escape). But I can tell this is a minor concern.

Given all this, do you think it makes sense to switch to `Option`? I kind of like how it is now, but it'll be fine either way.